### PR TITLE
[comgr][hotswap] speed up kernel-descriptor rewrites for large objects

### PR DESCRIPTION
**Pull request by nirmie**
## Summary

Three algorithmic fixes to the shared kernel-descriptor rewrite path. They are independent of the B0-to-B0 fast path but are exercised heavily by it, by the MC entry-trampoline path, and by the B0-to-A0 instruction-patch path on large (1000+ kernel) code objects.

- **`getKernelDescriptorVAddr` was O(n²).** It linearly scanned all descriptors, once per fixup (N fixups × N descriptors). Now O(1) via a name→vaddr `StringMap` built during descriptor-cache init.
- **`initializeKernelDescriptorCache` was O(n²).** It deduped each descriptor with an `std::any_of` linear scan over everything seen so far. Now O(n) via a seen-set `StringMap`.
- **`rewriteKernelEntryDescriptorOffsets` did an O(n²) msgpack round-trip.** It applied each kernel's SGPR-reservation bump with a per-kernel call that reparsed and reserialized the *entire* AMDGPU metadata note (which holds every kernel). Now all bumps are collected in the loop and applied in a single batched `updateKernelMetadataSgprCounts` call.

The batched-msgpack SGPR update was independently developed in parallel by @rahjain-amd in #3380; this PR lands the equivalent fix on the fast-path stack. #3380 Should be encompassed by this PR now. 

This is the perf half of a two-PR pair, stacked on the B0-to-B0 fast-path PR #3361. #3361 adds the fast path itself; this PR restores/adds the load-time optimizations that make it (and the MC path) fast on large objects. They are meant to land together.

## Performance

Offline full rewrite of the gfx1250 hipBLASLt monolith (`Kernels.so-000-gfx1250.hsaco`, 8.7 MB, ~1093 kernels), measured with the in-tree `hotswap-rewrite` tool:

| path | without these opts | with these opts |
|---|---|---|
| B0-to-B0 fast path | +33 ms over parity (large object) | at parity with the ROCr loader |
| per-kernel SGPR reserve (msgpack) | ~14.5 s | ~0.09 s |

The two O(n²) descriptor scans were the bulk of the fast path's residual ~33 ms gap to the ROCr loader on the 1093-kernel object; with them fixed the fast path reaches load-time parity. The msgpack batching is what makes any path that bumps per-kernel SGPR reservations (MC entry-trampoline, and the per-kernel-SGPR fast-path variant) usable on large objects at all, resulting in a 150x+ reduction.

All three are pure algorithmic changes (identical output, lower complexity), verified by the existing hotswap LIT suite (74 tests) and `--check-idempotent`.

## Testing

- `make check-comgr` hotswap LIT suite: all pass (unchanged output).
- Offline `hotswap-rewrite` on the hipBLASLt and hipSPARSELt gfx1250 monoliths: RESULT SUCCESS, idempotent.
- Running now on full suite of rocm-libraries tests on A0, will update for total coverage

## Reproducing the benchmark

The load-time harness lives in the internal rocm-hotswap-testing repo under `perf-harness/` (AMD-ROCm-Internal/rocm-hotswap-testing#17).

TLDR:

```bash
DIST=/path/to/rocm/dist ./perf-harness/run_hsa_load_comparison.sh results/
python3 perf-harness/analyze_hsa_load.py results/
```

It sweeps baseline / rocr_loader / comgr_hotswap / comgr_hotswap_fast cells over the gfx1250 monolith libraries and prints the per-object load floor plus the fast-vs-loader and MC-vs-loader deltas. See `perf-harness/README.md` there for the cell definitions, env knobs, and the B0-vs-A0 measurement caveat.

### CODE REVIEW

File: amd/comgr/src/comgr-hotswap-elf.cpp [hunk 1]
```diff
@@ -650,19 +651,30 @@ void ElfView::initializeKernelDescriptorCache() const {
...
-          });
-      if (!Seen)
+      StringRef KernelNameRef = NameOrErr->drop_back(3);
+      std::string KernelName = KernelNameRef.str();
+      // Dedup by (name, vaddr) via a map; a per-symbol O(n) scan made cache
+      // init O(n^2).
+      StringMap<uint64_t>::const_iterator SeenIt =
+          SeenVAddr.find(KernelNameRef);
+      const bool Seen =
+          SeenIt != SeenVAddr.end() && SeenIt->second == Sym.st_value;
```
Comment (harsh-amd) [inline amd/comgr/src/comgr-hotswap-elf.cpp]: This does not quite dedup by `(name, vaddr)` since the map only remembers the last vaddr for each name. For `A, B, A` under the same name, the last `A` is inserted again, unlike the old scan. Could we use a per-name set of vaddrs and add a test for this ordering?

File: amd/comgr/src/comgr-hotswap-entry-trampoline.cpp [hunk 2]
```diff
@@ -671,17 +675,16 @@ bool rewriteKernelEntryDescriptorOffsets(
...
-    // The fast path leaves the logical SGPR reservation unchanged (fixed
-    // s[100:101]), so skip the descriptor SGPR-count update entirely.
-    bool UpdatedSgprs = Fixup.SkipSgprReservation
-                            ? true
-                            : OutElf.updateKernelDescriptorSgprCount(
-                                  Fixup.KernelName, Fixup.RequiredSgprs,
-                                  /*UpdateDescriptor=*/false);
+    if (!Fixup.SkipSgprReservation && Fixup.RequiredSgprs != 0)
+      SgprBumps[Fixup.KernelName] =
+          std::max(SgprBumps.lookup(Fixup.KernelName), Fixup.RequiredSgprs);
```
Comment (harsh-amd) [inline amd/comgr/src/comgr-hotswap-entry-trampoline.cpp]: Could we add a multi-kernel MC-path test for this batching? The helper is covered, but I do not see coverage for the new max/skip/zero aggregation here or for propagating a missing-metadata failure.

**Comment by harsh-amd** [comment]
Since this PR is performance-focused, could we include the exact benchmark command, run count, median/dispersion, and peak memory? The current point estimates are difficult to reproduce, especially with the additional per-object maps.

File: amd/comgr/src/comgr-hotswap-elf.cpp (hunk 1 above)
Comment (nirmie) [reply amd/comgr/src/comgr-hotswap-elf.cpp]: Added, updated SeenVAddr to be a `StringMap<DenseSet<uint64_t>>`  to track kernel address alongside names to do true dedup. Added unit test too

File: amd/comgr/src/comgr-hotswap-entry-trampoline.cpp (hunk 2 above)
Comment (nirmie) [reply amd/comgr/src/comgr-hotswap-entry-trampoline.cpp]: Added two tests in HotswapMCTest.cpp that run
rewriteKernelEntryDescriptorOffsets end-to-end:

- AggregatesSgprBumpsMaxSkipZero: three kernels, each starting at sgpr_count = 8.
  The fixup list covers all three cases - one kernel appears twice (10 then 12,
  aggregates to the max), one sets SkipSgprReservation, one has RequiredSgprs = 0.
  Result: max kernel -> 12, the other two stay at 8.

- PropagatesMissingDescriptorFailure: a fixup for a kernel with no descriptor
  makes the whole call return false.

These need more than one kernel per object, so I added a makeMultiKernelDescriptorElf
builder in comgr-test-elf-utils.h (the existing single-kernel fixture is left
untouched).

**Comment by nirmie** [comment]
> Since this PR is performance-focused, could we include the exact benchmark command, run count, median/dispersion, and peak memory? The current point estimates are difficult to reproduce, especially with the additional per-object maps.

Going to post the testing harness in rocm-hotswap-testing, will describe it in the PR description

**Comment by lamb-j** [comment]
## Claude code review — #3382 (updated)

_Posted by @lamb-j (Claude Code). **Edited to drop points already raised by @harsh-amd** — the `(name, vaddr)` dedup keying and the multi-kernel batching test, both since addressed. Only my distinct findings remain below. Verified against head `58a35b21`._

Overall a clean set of algorithmic fixes that builds correctly on #3348's batching (metadata-only, no double-applied bump). Two small points:

### MINOR
- `comgr-hotswap-elf.cpp` (`getKernelDescriptorVAddr`) — separate from the dedup *set* that was fixed: the name->vaddr **lookup** map (`KernelDescriptorVAddrCache.try_emplace(name, vaddr)`) still collapses to first-vaddr-wins, so once the cache legitimately holds multiple vaddrs for a name, the lookup silently returns the first with no signal. Not a regression (the old linear scan also returned the first match in push order), but it contradicts the new "a name can map to >1 vaddr" comment. Reconcile — document first-in-symtab-wins, or `log()` when a name is seen with a divergent vaddr.

### NIT
- `entry-trampoline.cpp` — `SgprBumps[name] = max(lookup(...), ...)` does two hash lookups; `try_emplace` + conditional update halves it (negligible at these sizes).

Non-blocking coverage gap: no test for a kernel absent from the metadata note but present as a descriptor with a nonzero bump.

**Comment by nirmie** [comment]
> ### MINOR
> * `comgr-hotswap-elf.cpp` (`getKernelDescriptorVAddr`)

Updated the comment to be clear on the limitation that lookup deliberately keeps the first-in-symtab descriptor that the old code had, and its not for single-value name to vaddr resolution. Comment is in commit ea8adef8

> ### NIT
> * `entry-trampoline.cpp` — `SgprBumps[name] = max(lookup(...), ...)` does two hash lookups

Bound the mapped reference once and update in place, in commit ea8adef8

I can add the unit test in the next performance PR i mentioned in #3361

**Review by lamb-j — APPROVED** [review]
LGTM, thanks for the fixes
